cpp/src/arrow/c/abi.h

```diff
@@ -106,6 +212,98 @@ struct ArrowArrayStream {
 
 #endif  // ARROW_C_STREAM_INTERFACE
 
+#ifndef ARROW_C_DEVICE_STREAM_INTERFACE
+#define ARROW_C_DEVICE_STREAM_INTERFACE
+
+/// \brief Equivalent to ArrowArrayStream, but for ArrowDeviceArrays.
+///
+/// This stream is intended to provide a stream of data on a single
+/// device, if a producer wants data to be produced on multiple devices
+/// then multiple streams should be provided. One per device.
+struct ArrowDeviceArrayStream {
+  /// \brief The device that this stream produces data on.
+  ///
+  /// All ArrowDeviceArrays that are produced by this
+  /// stream should have the same device_type as set
+  /// here. The device_type needs to be provided here
+  /// so that consumers can provide the correct type
+  /// of queue_ptr when calling get_next.
+  ArrowDeviceType device_type;
+
+  /// \brief Callback to get the stream schema
+  /// (will be the same for all arrays in the stream).
+  ///
+  /// If successful, the ArrowSchema must be released independantly from the stream.
+  /// The schema should be accessible via CPU memory.
+  ///
+  /// \param[in] self The ArrowDeviceArrayStream object itself
+  /// \param[out] out C struct to export the schema to
+  /// \return 0 if successful, an `errno`-compatible error code otherwise.
+  int (*get_schema)(struct ArrowDeviceArrayStream* self, struct ArrowSchema* out);
+
+  /// \brief Callback to get the device id for the next array.
+  ///
+  /// This is necessary so that the proper/correct stream pointer can be provided
+  /// to get_next.
+  ///
+  /// The next call to `get_next` should provide an ArrowDeviceArray whose
+  /// device_id matches what is provided here, and whose device_type is the
+  /// same as the device_type member of this stream.
+  ///
+  /// \param[in] self The ArrowDeviceArrayStream object itself
+  /// \param[out] out_device_id Pointer to be populated with the device id, must not be
+  /// null \return 0 if successful, an `errno`-compatible error code otherwise.
+  int (*get_next_device_id)(struct ArrowDeviceArrayStream* self, int64_t* out_device_id);
+
+  /// \brief Callback to get the next array
+  ///
+  /// If there is no error and the returned array has been released, the stream
+  /// has ended. If successful, the ArrowArray must be released independently
+  /// from the stream.
+  ///
+  /// Because different frameworks use different types to represent this, we
+  /// accept a void* which should then be reinterpreted into whatever the
+  /// appropriate type is (e.g. cudaStream_t) for use by the producer.
+  ///
+  /// \param[in] self The ArrowDeviceArrayStream object itself
+  /// \param[in] queue_ptr The appropriate queue, stream, or
+  /// equivalent object for the device that the data is allocated on
+  /// to indicate where the consumer wants the data to be accessible.
+  /// If queue_ptr is NULL then the default stream (e.g. CUDA stream 0)
+  /// should be used to ensure that the memory is accessible from any stream.
```

Comment:
I'm a little confused here.  It sounds like I need to call `get_next_device_id` to determine which queue to use and then I need to pass that queue on to the call to `get_next`.  But why?  Why isn't the producer managing the queues?

If the producer controls which device id gets used (`get_next_device_id` seems to suggest this) then why does the consumer need to give it the queue?  For example, if I were merging streams from two different devices it seems like I would do something like (apologies for the butchered pseudo-code)...

```
// Dummy class merging two infinite streams in an inefficient round-robin fashion
class MergedStream {

  int get_next(ArrowDeviceArray* out) {
    if (merged_arrays_.empty()) {
      ArrowDeviceArray arr;
      left_.get_next(&arr);
      merged_arrays_.push(arr);
      right_.get_next(&arr);
      merged_arrays_.push(arr);
    }
    *out = merged_arrays_.pop();
  }
};
```

Reply:
@pitrou the issue is that there isn't a mechanism that you could call in the release callback (to my knowledge) to cleanly control the lifetime. (@kkraus14 correct me if I'm wrong and this isn't what you meant)

Reply:
That doesn't make sense, does it? How is the consumer supposed to manage the stream's lifetime if "there isn't a mechanism that you could call in the release callback to cleanly control the lifetime"?

Reply:
The idea of a consumer providing a stream is that it only needs to manage the lifetime of its provided stream. A flow would look something like:

- Producer creates stream 3
- Producer runs some kernel that produces data safe to consume on stream 3
- Consumer creates stream 4
- Consumer calls Producer API asking for C Data Interface and provides it stream 4
- Producer creates an Event (see cudaEventCreate or hipEventCreate) and records it on stream 3
- Producer waits the passed in stream 4 on the event (see cudaStreamWaitEvent or hipStreamWaitEvent)
- Producer returns the C Data Interface struct to the consumer

With this, the producer's stream 3 can be destroyed at any point safely after the C Data Interface struct is returned so you don't need to worry about two different libraries / frameworks needing to be able to control the lifetimes of the others' streams.

Reply:
Ok, what's wrong with the following:
* Producer creates CUDA stream 3
* Producer runs some kernel that produces data safe to consume on CUDA stream 3
* Consumer gets a C Data Interface export of the previously produced data; the `ArrowArray`'s release callback is responsible for handling the lifetime of stream 3, in addition to the lifetime of the allocated data
* Consumer either syncs on stream 3 directly, or chooses to sync stream 3 to its own stream 4 using the event APIs
* When consumer is done, it calls the release callback, letting the producer dealing with CUDA stream 3 lifetime

Reply:
Ok, so Tensorflow doesn't expose its CUDA streams to the user... but can the user request that Tensorflow data be tied to a specific CUDA stream? Otherwise, how can we implement the "consumer-provided CUDA stream" semantics?

Reply:
> can the user request that Tensorflow data be tied to a specific CUDA stream?

No a user can't request that.

> Otherwise, how can we implement the "consumer-provided CUDA stream" semantics?

Tensorflow can handle the consumer provided stream internally as part of it's implementation in producing dlpack or Arrow C Data Interface structs.

Reply:
Why wouldn't they? They can easily refcount the usage of their own CUDA streams.

Reply:
> Why wouldn't they? They can easily refcount the usage of their own CUDA streams.

I think that is making a lot of assumptions about how folks use and manage CUDA streams 😄. Again, some places use them similarly to thread pools and only control the lifetime of the pool.

I tried to dig through Tensorflow's code to figure exactly how they're managing the lifetime of their streams but I'm not confident, everything I say below may not be correct:

- Something eventually calls down to `AllocateStream` and `DeallocateStream` (https://github.com/tensorflow/tensorflow/blob/b9fc6a9b611ec373c02e5b5ab432b1d7aff9392e/tensorflow/compiler/xla/stream_executor/cuda/cuda_gpu_executor.cc#L759-L774) to create and destroy CUDA streams.
- These operate on raw ptrs and it looks like there's a class that wraps these, `Stream` (https://github.com/tensorflow/tensorflow/blob/b9fc6a9b611ec373c02e5b5ab432b1d7aff9392e/tensorflow/compiler/xla/stream_executor/stream.cc#L262-L286) which has constructor, destructor, and init functions roughly of what you'd expect.
- I believe these `Stream` objects are managed in a `StreamPool` (https://github.com/tensorflow/tensorflow/blob/b9fc6a9b611ec373c02e5b5ab432b1d7aff9392e/tensorflow/compiler/xla/service/stream_pool.h#L27-L59) which then allows "borrowing" the streams using unique ptrs.

I guess in theory that if they ultimately have `Stream` objects being used that it could be moved into the private data being used by the release callback.

Reply:
Okay, regardless if we take the producer provided path then I think it makes a lot more sense for the producer to share an Event than a Stream.

An Event can be waited on via `cudaStreamWaitEvent` / `hipStreamWaitEvent` which does a device side wait which would have minimal overhead if it's the same stream or `cudaEventSynchronize` / `hipEventSynchronize` if blocking host code is desired.

Reply:
Since it seems we're going to take the Producer providing an event path, there isn't really a need for the `get_next_device_id` callback anymore, correct? Or am I missing something?